examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3181 @@
+    ) -> ContractResult<bool> {
+        // Setup parameter.
+        let parameter_bytes = to_bytes(&IsOperatorParams {
+            owner:   *owner,
+            address: *address,
+        });
+
+        let state_address = host
+            .state()
+            .state_address
+            .ok_or(concordium_cis2::Cis2Error::Custom(CustomContractError::StateInvokeError))?;
+
+        let is_operator = host.invoke_contract_raw_read_only(
+            &state_address,
+            Parameter(&parameter_bytes),
+            EntrypointName::new_unchecked("isOperator"),
+            Amount::zero(),
+        )?;
+
+        // It is expected that this contract is initialized with the w_ccd_state
+        // contract (a V1 contract). In that case, the balance variable can be
+        // queried from the state contract without error.
+        let is_operator = is_operator
+            .ok_or(concordium_cis2::Cis2Error::Custom(CustomContractError::StateInvokeError))?
+            .get()?;
+
+        Ok(is_operator)
+    }
+}
+
+// Contract functions
+
+/// Initialize the state contract instance with no initial tokens.
+/// Logs a `Mint` event for the single token id with no amounts.
+#[init(contract = "CIS2-wCCD-State", enable_logger)]
+fn contract_state_init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+    logger: &mut impl HasLogger,
+) -> InitResult<State<S>> {
+    // Construct the initial contract state.
+    let state = State::new(state_builder);
+    // Get the instantiater of this contract instance.
+    let invoker = Address::Account(ctx.init_origin());
+    // Log event for the newly minted token.
+    logger.log(&Cis2Event::Mint(MintEvent {
+        token_id: TOKEN_ID_WCCD,
+        amount:   ContractTokenAmount::from(0u64),
+        owner:    invoker,
+    }))?;
+
+    // Log event for where to find metadata for the token
+    logger.log(&Cis2Event::TokenMetadata::<_, ContractTokenAmount>(TokenMetadataEvent {
+        token_id:     TOKEN_ID_WCCD,
+        metadata_url: MetadataUrl {
+            url:  String::from(TOKEN_METADATA_URL),
+            hash: None,
+        },
+    }))?;
+
+    Ok(state)
+}
+
+/// Initialize contract instance with no initial tokens.
+/// Logs a `Mint` event for the single token id with no amounts.
+#[init(contract = "CIS2-wCCD", enable_logger)]
+fn contract_init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    _state_builder: &mut StateBuilder<S>,
+    logger: &mut impl HasLogger,
+) -> InitResult<StateImplementation> {
+    // Get the instantiater of this contract instance.
+    let invoker = Address::Account(ctx.init_origin());
+    // Construct the initial contract state.
+    let state = StateImplementation::new(invoker, invoker);
+    // Log event for the newly minted token.
+    logger.log(&Cis2Event::Mint(MintEvent {
+        token_id: TOKEN_ID_WCCD,
+        amount:   ContractTokenAmount::from(0u64),
+        owner:    invoker,
+    }))?;
```

Comment:
It is not required by the standard to do this, but the purpose here is to announce the existence of the token and the token id.
The total supply at this point is 0, but the token does exist.

Reply:
Would it be better to use the following event here:

Cis2TokenDeployed{
token_address: ContractAddress
token_id: TOKEN_ID_WCCD,
owner:    invoker,
}

Reply:
You mean introduce an event for logging the existence of a token, and another for minting of this token?
It would make it a bit more explicit, but for tokens which are not starting out with a total supply of 0, it would mean having to log more events and I don't think it would provide much value.

Reply:
Yes, if minting is happening during the deployment of a token then we need to emit a `MintEvent` in addition to the `Cis2TokenDeployedEvent`.

The advantage of adding a separate `Cis2TokenDeployedEven` is that off-chain indexers can insert the `token_address` in their off-chain database easily without having to skim to `MintEvents` to find a `Mint` event that is associated with a new token and then have to query to token_address or look it up from the tx.receipt data.

Reply:
I added this comment above the the `mint` event:

// Log a `Mint` event for the single token id with 0 amount. Minting an amount
// of 0 is used to signal that a new token was deployed. Invoking the mint event
// with 0 amount in the `wrap` function is impossible.